File: client/common/bet.go
package common

import (
	"fmt"
	"strconv"
	"strings"
	"net"
)

type Bet struct {
	Agency     uint8
	FirstName string
	LastName  string
	Document   uint32
	Birthdate  string
	Number     uint32
}

func sendBet(b Bet, conn net.Conn)  error {
	var data []byte

	agency := b.Agency
	data = append(data, byte(agency))

	firstNameBytes := []byte(b.FirstName)
	data = append(data, byte(len(firstNameBytes)))
	data = append(data, firstNameBytes...)

	lastNameBytes := []byte(b.LastName)
	data = append(data, byte(len(lastNameBytes)))
	data = append(data, lastNameBytes...)

	data = append(data, byte(b.Document>>24), byte(b.Document>>16), byte(b.Document>>8), byte(b.Document))

	dateParts := strings.Split(b.Birthdate, "-")
	year, _ := strconv.Atoi(dateParts[0])
	month, _ := strconv.Atoi(dateParts[1])
	day, _ := strconv.Atoi(dateParts[2])

	data = append(data, byte(year>>8), byte(year))
	data = append(data, byte(month))
	data = append(data, byte(day))

	data = append(data, byte(b.Number>>24), byte(b.Number>>16), byte(b.Number>>8), byte(b.Number))

	err := safeWrite(conn, append(data, byte('|')))

	if err != nil {
		return fmt.Errorf("failed to send the bet: %v", err)
	}

	return nil
}

func receiveConfirm(conn net.Conn, bet Bet) error {
	data, err := safeRead(conn, 2)

	if err != nil {
		return fmt.Errorf("failed to receive the server response: %v", err)
	}

	msg := string(data)

	if msg != "OK" {
		return fmt.Errorf("failed to receive the server response: %v", err)
	}
	return nil
}

func safeWrite(conn net.Conn, bytes []byte) error {
	totalBytesWritten := 0
	for totalBytesWritten < len(bytes) {
		bytesWritten, err := conn.Write(bytes[totalBytesWritten:])

		if err != nil || bytesWritten == 0 {
			return fmt.Errorf("Error sending the batch: %v", err)
		}

		totalBytesWritten += bytesWritten
	}
	
	return nil
}

func safeRead(conn net.Conn, sizeToRead int) ([]byte, error) {
	totalBytesRead := 0
	data := make([]byte, sizeToRead)

	for totalBytesRead < sizeToRead {
		bytesRead, err := conn.Read(data[totalBytesRead:])

		if err != nil || bytesRead == 0 {
			return nil, fmt.Errorf("Error reading from the server: %v", err)
		}

		totalBytesRead += bytesRead
	}

	return data, nil
}

File: server/common/bet.py
    
    

import logging
from common.utils import Bet


def readBetFromBytes(client_sock):
    data = safeRead(client_sock, 1024)

    bet = parseBetFromBytes(data)

    return bet


def parseBetFromBytes(data):
    """Deserializes a byte array to a Bet object."""

    index = 0

    agency = int.from_bytes(data[index:index + 1], byteorder='big')
    index += 1

    first_name_length = data[index]
    index += 1
    first_name = data[index:index + first_name_length].decode('utf-8')
    index += first_name_length
    
    last_name_length = data[index]
    index += 1
    last_name = data[index:index + last_name_length].decode('utf-8')
    index += last_name_length

    dni = int.from_bytes(data[index:index + 4], byteorder='big')
    index += 4

    year = int.from_bytes(data[index:index + 2], byteorder='big')
    index += 2
    month = int.from_bytes(data[index:index + 1], byteorder='big')
    index += 1
    day = int.from_bytes(data[index:index + 1], byteorder='big')
    index += 1

    birth_date = f"{year}-{month:02d}-{day:02d}"

    number = int.from_bytes(data[index:index + 4], byteorder='big')
    
    return Bet(agency,first_name, last_name, str(dni), birth_date, str(number))

def confirmBet(client_sock):
    safeWrite(client_sock, b"OK")

def safeRead(client_sock, amount):
    """Safely reads a specified amount of data from the client socket."""
    buffer = b""

    while len(buffer) < amount and not buffer.endswith(b'|'):
        chunk = client_sock.recv(amount)
        if not chunk:
            raise Exception("Connection closed by the client")
        buffer += chunk

    return buffer

def safeWrite(client_sock, bytes):
    """Safely writes data to the client socket."""
    totalBytesWritten = 0
    dataLength = len(bytes)

    while totalBytesWritten < dataLength:
        try:
            bytesWritten = client_sock.send(bytes[totalBytesWritten:])
            if bytesWritten == 0:
                raise Exception("Connection closed by the client")
            totalBytesWritten += bytesWritten
        except Exception as e:
            raise Exception(f"Error sending the batch: {e}")

    return totalBytesWritten